scipy/integrate/_quadrature.py

```diff
@@ -822,6 +824,294 @@ def simpson(y, *, x=None, dx=1.0, axis=-1, even=_NoValue):
     return result
 
 
+def _cumulatively_sum_simpson_subintegrals(
+    sub_integrals_h1: np.ndarray,
+    sub_integrals_h2: np.ndarray,
+) -> np.ndarray:
+    """Calculate cumulative sum of Simpson subintegrals.
+
+    Takes as input the estimated Simpson integrals over h1 and
+    h2 subintervals. Returns the cumulative sum using
+    composite Simpson's rule. Assumes the axis of summation is 0.
+    """
+    # Set integral of odd h1 intervals = 0
+    sub_integrals_h1[1::2] = 0.0
+    # Set integral of odd h2 intervals = 0, except for the last interval
+    sub_integrals_h2[1:-1:2] = 0.0
+    sub_integrals = np.concatenate(
+        [
+            # Integral over first subinterval can only be calculated from 
+            # formula for h1
+            sub_integrals_h1[0:1],
+            # Combine sub_integral arrays (at every index, i, either
+            # sub_integrals_h1[i] = 0 or sub_integrals_h2[i] = 0)
+            (sub_integrals_h1[1:] + sub_integrals_h2[:-1]),
+            # Integral over last subinterval can only be calculated from 
+            # formula for h2
+            sub_integrals_h2[-1:],
+        ],
+        axis=0,
+    )
+    res = np.cumsum(sub_integrals, axis=0)
+    return res
+
+
+def _cumulative_simpson_equal_intervals(
+    y: np.ndarray, dx: np.ndarray, axis: int
+) -> np.ndarray:
+
+    y = y.swapaxes(0, axis)
+    dx = dx.swapaxes(0, axis)
+
+    # Consider a quadratic interpolation over each set of 3 adjacent points,
+    # a, a+h1, a+h1+h2. The subinterval widths are h1 and h2
+    # (in this case, h1 = h2 = dx)
+
+    # Calculate integral over the h1 subintervals.
+    # Calculate for all but last subinterval of y.
+    sub_integrals_h1 = dx[:-1] / 3.0 * \
+        (5.0 * y[:-2] / 4.0 + 2.0 * y[1:-1] - y[2:] / 4.0)
```

Comment:
These are easy to fit on one line by omitting the `0`s. TBH, I think you can omit the `.0`s.

Reply:
Updated these to use parentheses.
The .0s ensure conversion to floats. There is an edge case where if the inputs are int arrays, the integration output may end up as an int array as well giving inaccurate results. Having the .0s ensures the result returns an accurate result with a float array.

Reply:
You can still omit the `0`s. 

I'd suggest ensuring float dtype from the outset. It avoids such edge cases, and it doesn't require repeated promotion of the whole array to float before each operation.

```
import numpy as np
x = np.arange(999999)
%timeit x*5.
# 3.13 ms ± 346 µs per loop
x = x.astype(float)
%timeit x*5
# 1.43 ms ± 443 µs per loop
```

Reply:
I have removed the .0s and fit the lines you highlighted to a single line.
Regarding casting to float, earlier @MatteoRaso raised a valid concern; what if the input is float128? If we cast astype(float), the array will be converted to float64 which is lower precision than the input (https://github.com/scipy/scipy/pull/18151#discussion_r1167135755).

To address this, I am now only casting to float if the input is an int array (https://github.com/scipy/scipy/pull/18151/files#diff-2ff88b8913ad74b9ea3fe1fe02d6d49164855189168b767076af2e94206c3a2cR938).

Reply:
The purpose of my code was to demonstrate the advantage of casting to a floating point type. I didn't mean to cast to `float` unconditionally. Using `issubdtype` to detect integers is exactly what I would have recommended, so this looks good.